### PR TITLE
build: bring back ts-optchain's oc

### LINKED ISSUE #282: Bring back ts-optchain's oc
In the past, we used [ts-optchain](https://github.com/rimeto/ts-optchain)'s oc, but we found a bug (it cannot refer a type of properties or sub properties in some specific cases), so we fixed it [here](https://github.com/crazyfactory/ts-react-boilerplate/blob/master/src/app/helpers/oc.ts) but never sent a pr. Apparently that bug which I don't remember has been fixed, so we can bring it back.

### PR DESCRIPTION
**Pull request by sanketmaru**
Added ts-optchain to fix #282 as a dependency.

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -12,7 +12,10 @@
     "noUnusedLocals": true,
     "jsx": "react",
     "moduleResolution": "node",
-    "esModuleInterop": true
+    "esModuleInterop": true,
+    "plugins": [
+      { "transform": "ts-optchain/transform" },
```
Comment (paibamboo) [inline tsconfig.json]: image


This doesn't work yet (still use `oc` and not transpiling the code). To make it work I think you need to install [ttypescript](https://github.com/cevek/ttypescript), and move this transform config to webpack config instead (Not sure why it doesn't work when this config is in `tsconfig.json`).

Anyway, I'm fine if you just remove this part of code, and live without the transpiled code for now. We didn't support IE11 with the old `oc` anyway because it uses Proxies, and Proxies can't be polyfilled.

Also, I think we will migrate from `ts-loader` to `babel-loader` soon. So after that we could use [babel-plugin](https://github.com/epeli/babel-plugin-ts-optchain) for transpiling the code instead.

In summary either
1) make this `transform` plugin work, or
2) remove this `transform` plugin config for now

Comment (sanketmaru) [reply tsconfig.json]: I will check it again and make the appropriate changes.

Comment (sanketmaru) [reply tsconfig.json]: I used ttypescript and made the changes in webpack config in dev.js 
```js
{
        test: /\.tsx?$/,
        loader: 'ts-loader',
        options: {
          compiler: 'ttypescript'
        },
        exclude: /node_modules/
}
```
No change, i can see this code ```js exports.translationsSelector = function (state) { return ts_optchain_1.oc(state).settings.translations(); }; ``` and is not transpiled. 

Any guess why its not doing so ?

Comment (paibamboo) [reply tsconfig.json]: Require `ts-optchain/transform`
```
const optChainTransform = require('ts-optchain/transform').default;
```

And try adding `getCustomTransformers` to `options` like so:
```
options: {
  compiler: 'ttypescript',
  getCustomTransformers: (program) => ({
    after: [optChainTransform(program)],
  })
}
```

Comment (sanketmaru) [reply tsconfig.json]: Thanks @paibamboo This works. I have pushed the changes for dev and prod.

Comment (cyberhck) [reply tsconfig.json]: can you check also for server config? that's used for server side rendering

**Comment by cyberhck** [comment]
hey, does it now transpile or use the normal proxy way like it used to?
if it transpile so the result bundle doesn't contain oc, maybe it's a good idea to remove it from dependency and add it to devDependency? I missed this PR here,

**Comment by sanketmaru** [comment]
@cyberhck , yes it does now transpile and result bundle doesn't contain oc. 
Will raise PR for adding oc as a devDependency.
